modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+    a severe imbalance, or when you have a smaller number of elements to balance.
+    Furthermore, while this operation is parallel-safe, it should be called in a
+    sequential context.
+  */
+  proc balance() {
+    var localThis = getPrivatizedThis;
+    // Phase 1: Acquire all locks from from first node and segment to last
+    // node and segment (our global locking order...)
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.acquire(STATUS_BALANCE);
+      }
+    }
+
+    // Phase 2: Concurrently redistribute elements from segments which contain
+    // more than the computed average.
+    coforall segmentIdx in 0 .. #here.maxTaskPar {
+      var nSegmentElems : [localThis.targetLocales.size] int;
+      var locIdx = 0;
+      for loc in localThis.targetLocales do on loc {
+        var nElems = getPrivatizedThis.bag.segments[segmentIdx].nElems.read() : int;
+        nSegmentElems[locIdx] = nElems;
+        locIdx += 1;
+      }
+
+      // Find the average and the excess. The excess is calculated as the amount
+      // of elements a segment has over the average, which is used to calculate
+      // the buffer size for each segment.
+      var total = (+ reduce nSegmentElems);
+      var avg = total / locIdx;
+      var excess : int;
+      for nElems in nSegmentElems {
+        if nElems > avg {
+          excess += nElems - avg;
+        }
+      }
+
+      // Allocate buffer, which holds the 'excess' elements for redistribution.
+      // Then fill it.
+      var buffer = c_malloc(eltType, excess);
+      var bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if nElems > average {
+          var take = nElems - average;
+          var tmpBuffer = buffer + bufferOffset;
+          segment.transferElements(tmpBuffer, take, buffer.locale.id);
+          bufferOffset += take;
+        }
+      }
+
+      // With the excess elements, redistribute it...
+      // BUG: Need to use Chapel Arrays due to issues with compiler automatically
+      // dereferencing a C_Ptr resulting in a segfault.
+      // BUG: Without '--no-denormalization' flag, the compiler will incorrectly
+      // use everything beyond this point by reference and *not* dereference it
+      // beforehand, causing more undefined behavior. For now, I use Chapel Arrays.
+      bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if average > nElems {
+          var give = average - nElems;
+          var arr : [1..give] eltType;
+          on bufferOffset {

```

Comment:
We call `on bufferOffset` because bufferOffset is located on the node that holds the `buffer`, which is the node that called `balance`. On line 341, you see that we do a `for loc in Locales do on loc`. We have to jump to the locale owning the `buffer`  because its a C_Ptr.